### ISSUE TITLE
npm install csound-api not working

### ISSUE THREAD

**Issue by xenophoria** [nwhetsell/csound-api#18]
I am trying to upgrade the current version of QTCsound on a MacBook Pro (recently upgraded to Catalina 10.15.1) so I installed **Homebrew** before installing **Boost** and **Csound** in Terminal following step one, i.e. **brew install boost csound**. When I try to install the package in step 2 **npm install csound-api** the following error message appears in Terminal

```
> csound-api@0.10.16 install /Users/gs/node_modules/csound-api
> node-gyp rebuild

  CXX(target) Release/obj.target/csound-api/src/csound-api.o
../src/csound-api.cc:3:10: fatal error: 'csound/cwindow.h' file not found

stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
```

Here are the first ten lines of the log /Users/gs/.npm/_logs/2019-12-10T07_55_15_791Z-debug.log 

```
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   '/usr/local/Cellar/node/12.5.0/bin/node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'install',
1 verbose cli   'csound-api'
1 verbose cli ]
2 info using npm@6.9.0
3 info using node@v12.5.0
4 verbose config Skipping project config: /Users/gs/.npmrc. (matches userconfig)
```

Help ! What should I do ?

**Comment by nwhetsell** [comment]
What’s the output when you run in Terminal `brew reinstall csound`?

**Comment by xenophoria** [comment]
This complains about Xcode being too outdated.

```
Gregorys-MacBook-Pro-2:~ gs$ brew reinstall csound
==> Reinstalling csound 
==> Installing dependencies for csound: llvm, faust, fltk, gdbm, readline, sqlite, xz, python, glib, portaudio, fluid-synth, isl, mpfr, libmpc, gcc, szip, hdf5, berkeley-db, libsamplerate, jack, liblo, openblas, numpy, portmidi, stk and wiiuse
==> Installing csound dependency: llvm
Error: Your Xcode (10.2.1) is too outdated.
Please update to Xcode 11.2.1 (or delete it).
Xcode can be updated from the App Store.

Gregorys-MacBook-Pro-2:~ gs$
```

I'm not sure why this is relevant. I was hoping to use QTCsound to create HTML as front end for Csound.

**Comment by xenophoria** [comment]
I'll update Xcode to 11.2.1 and try this again

**Comment by nwhetsell** [comment]
> I'm not sure why this is relevant.

It’s relevant because it means the installation of Csound failed (and consequently csound/cwindow.h is not available).

> I'll update Xcode to 11.2.1 and try this again

Please do. Updating Xcode (and then running `brew reinstall csound` again) should let you install csound-api.

**Comment by xenophoria** [comment]
I upgraded to **Xcode 11.3** which took care of `llvm` dependencies. I see the relevance now, thanks. `brew reinstall csound` worked (a massive script, thanks for your time). Csound installation is updated but the same error report happens when I run `npm install csound-api` in the Terminal.

```
CXX(target) Release/obj.target/csound-api/src/csound-api.o
../src/csound-api.cc:3:10: fatal error: 'csound/cwindow.h' file not found
#include <csound/cwindow.h>
         ^~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [Release/obj.target/csound-api/src/csound-api.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)

```

**Comment by nwhetsell** [comment]
What’s the output of running `brew reinstall csound` again? (The installation should be much faster this time.)

**Comment by xenophoria** [comment]
Yes, only a minute this time. There is one mystery. 

```
==> `brew cleanup` has not been run in 30 days, running now...
Error: Permission denied @ apply2files - /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuProN-W3
Gregorys-MacBook-Pro-2:~ gs$ 
```

I tried to delete `/usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuProN-W3` but when I try to remove it (with `rm`) Terminal reports **Permission denied**. As I recall `HiraKakuProN-W3` was something I used when I experimented with drawing to a canvas in javascript. But I don't need it so I'm happy to get rid of it. How ?

**Comment by xenophoria** [comment]
What's this !!!

Gregorys-MacBook-Pro-2:~ gs$ ls -l /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuProN-W3
lrwxr-xr-x  1 gs  wheel  58 25 Dec  2015 /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuProN-W3 -> /System/Library/Fonts/ヒラギノ角ゴ ProN W3.otf

**Comment by nwhetsell** [comment]
The output of running `brew reinstall csound` that I’m expecting to see should look like this:

```
2018-MBP% brew reinstall csound
==> Reinstalling csound 
==> Downloading https://homebrew.bintray.com/bottles/csound-6.13.0_3.mojave.bott
Already downloaded: /Users/nathanwhetsell/Library/Caches/Homebrew/downloads/4170fc006e6445b5ba07f096d03ddd8ef334132dc0436c080d7237df64d14377--csound-6.13.0_3.mojave.bottle.1.tar.gz
==> Pouring csound-6.13.0_3.mojave.bottle.1.tar.gz
Warning: csound dependency gcc was built with a different C++ standard
library (libstdc++ from clang). This may cause problems at runtime.
==> Caveats
To use the Python bindings, you may need to add to ~/.bash_profile:
  export DYLD_FRAMEWORK_PATH="$DYLD_FRAMEWORK_PATH:/usr/local/opt/csound/Frameworks"
==> Summary
🍺  /usr/local/Cellar/csound/6.13.0_3: 135 files, 6.2MB
2018-MBP% 
```

There’s almost no explanation for Clang not being able to find csound/cwindow.h besides something awry with Csound’s installation, so I need to see the output of `brew reinstall csound`.

The `brew cleanup` error is almost certainly unrelated, and I’m unable to assist with that one.

**Comment by xenophoria** [comment]
Here's the output of` brew install csound`
[brew install csound Output.txt](https://github.com/nwhetsell/csound-api/files/3947548/brew.install.csound.Output.txt)

**Comment by xenophoria** [comment]
Keep in mind I'm running Catalina 10.15.1. I notice your last message had
`==> Pouring csound-6.13.0_3.mojave.bottle.1.tar.gz`

**Comment by xenophoria** [comment]
and also
```
Gregorys-MacBook-Pro-2:~ gs$ ls /usr/local/Cellar/csound
6.13.0_3
```

**Comment by nwhetsell** [comment]
Where it says—

```
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink bin/atsa
Target /usr/local/bin/atsa
already exists. You may want to remove it:
  rm '/usr/local/bin/atsa'

To force the link and overwrite all conflicting files:
  brew link --overwrite csound

To list all files that would be deleted:
  brew link --overwrite --dry-run csound
```

—that’s important. A symlink from /usr/local/include/csound to the actual location of Csound’s headers isn’t created if `brew link` is unsuccessful. You should do what the error message says, specifically—

```
To force the link and overwrite all conflicting files:
  brew link --overwrite csound

To list all files that would be deleted:
  brew link --overwrite --dry-run csound
```

—and then you should be able to install csound-api.

csound-api’s tests use GitHub Actions, which runs Catalina. The [tests are currently passing](https://github.com/nwhetsell/csound-api/runs/315854526), so it’s a certainty that csound-api can be installed on that platform.

**Comment by xenophoria** [comment]
I created the list of would be deletions ` brew link --overwrite --dry-run csound`

(output deletions)

[Would be deletion.txt](https://github.com/nwhetsell/csound-api/files/3947615/Would.be.deletion.txt)

then 

```
Gregorys-MacBook-Pro-2:~ gs$ brew link --overwrite csound
Linking /usr/local/Cellar/csound/6.13.0_3... 2635 symlinks created
Gregorys-MacBook-Pro-2:~ gs$ 
```

Looking promising, there are a few warnings but no errors
```
Gregorys-MacBook-Pro-2:~ gs$ npm install csound-api

> csound-api@0.10.16 install /Users/gs/node_modules/csound-api
> node-gyp rebuild

  CXX(target) Release/obj.target/csound-api/src/csound-api.o
  SOLINK_MODULE(target) Release/csound-api.node
npm WARN saveError ENOENT: no such file or directory, open '/Users/gs/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/Users/gs/package.json'
npm WARN gs No description
npm WARN gs No repository field.
npm WARN gs No README data
npm WARN gs No license field.

+ csound-api@0.10.16
added 2 packages from 9 contributors and audited 56 packages in 4.4s
found 0 vulnerabilities

Gregorys-MacBook-Pro-2:~ gs$ 
```
I will test later today and report. I owe you a beer at least. Thanks.

**Comment by nwhetsell** [comment]
The warnings are pretty much expected—this is `npm` telling you that you’re installing csound-api in your home directory (/Users/gs) instead of a directory containing a Node.js package. Your home directory has no description, no repository field, no README data, and no license field (see also https://github.com/nwhetsell/csound-api/issues/17#issuecomment-499096988).

**Comment by xenophoria** [comment]
Following yesterday's cleanup I finally tried running CsoundQt using Csound API (i.e. by opening **CsoundQt Configuration** and selecting `Run Utilities using: Csound API`)

In the Csound Output Console I get the following message

**unknown CSD tag: <html> Hello, World, this is Csound!</html>**

Un-highlighted syntax in the CSD file also shows `<html>` is not recognised as a CSD tag.


<img width="348" alt="testCSD" src="https://user-images.githubusercontent.com/19865210/70684562-a393de00-1cfa-11ea-9396-c9012cd3fba9.png">


As I tried to reinstall `csound-api` this notification appeared in Terminal

```
New minor version of npm available! 6.9.0 → 6.13.4       │
   │   Changelog: https://github.com/npm/cli/releases/tag/v6.13.4   │
   │               Run npm install -g npm to update!

```
I updated `npm` and tried to reinstall `csound-api` by typing `npm install csound-api` in Terminal


<img width="450" alt="Screen Shot 2019-12-12 at 4 22 29 pm" src="https://user-images.githubusercontent.com/19865210/70684847-ae9b3e00-1cfb-11ea-9183-efc805f59e12.png">


This indicates `csound-api` was installed but when I tried finding it I got this:

```
Gregorys-MacBook-Pro-2:bin gs$ which csound-api
Gregorys-MacBook-Pro-2:bin gs$ 
```
My questions are : 
where is `csound-api` ? 
and how do I get `CsoundQt` to find it ? 
and recognise the CSD tag `<html>` ?

**Comment by nwhetsell** [comment]
As it says in [csound-api’s README](https://github.com/nwhetsell/csound-api#readme), csound-api is a [Node.js Addon](https://nodejs.org/api/addons.html) for using Csound through its C [API](https://csound.com/docs/api/). As far as I know, csound-api is totally independent of [CsoundQt](https://github.com/CsoundQt/CsoundQt). If you’re trying to write JavaScript code to run Csound in [Node.js](https://nodejs.org/en/about/) (examples [in the README](https://github.com/nwhetsell/csound-api#examples) and the [examples folder](https://github.com/nwhetsell/csound-api/tree/master/examples)), then csound-api is one way to do that. Otherwise, you probably don’t need csound-api.

To answer your questions:

* csound-api is located where you installed it. The output when you ran `npm install csound-api` says that’s /usr/local/bin/node_modules/csound-api. (Running `which csound-api` would produce no output even if csound-api was located in /usr/local/bin since csound-api is a folder containing a Node.js Addon, not an executable.)

* I don’t use CsoundQt, so I don’t know how to configure it. You may be able to get help at https://github.com/CsoundQt/CsoundQt/issues, but I highly doubt that CsoundQt integrates with Node.js since the CsoundQt repository [contains no JavaScript code](https://github.com/CsoundQt/CsoundQt/search?l=javascript).

* What you have between `<html>` and `</html>` isn’t a valid webpage, so I wouldn't expect it to work, but if you think Csound is outputting unexpected messages about CSD tags, you may be able to get help at https://listserv.heanet.ie/cgi-bin/wa?A0=CSOUND. You may also want to look at the [example files included in CsoundQt](https://github.com/CsoundQt/CsoundQt/tree/0816047d7fb48e7d2f7ba88cd484a240c00d8528/examples/Html5%20Support). If CsoundQt’s syntax-highlighting looks unexpected, you can create a new issue at https://github.com/CsoundQt/CsoundQt/issues.

I’m closing this issue since csound-api appears to be installing as expected. If you have any more issues with csound-api (not CsoundQt or Csound itself), feel free to open a new issue.
